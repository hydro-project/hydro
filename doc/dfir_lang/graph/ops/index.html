<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="DFIR’s operators"><title>dfir_lang::graph::ops - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../../static.files/rustdoc-e56847b5.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="dfir_lang" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (f6092f224 2025-09-22)" data-channel="nightly" data-search-js="search-e256b49e.js" data-stringdex-js="stringdex-061df703.js" data-settings-js="settings-c38705f0.js" ><script src="../../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-6dc2a7f3.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module ops</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../dfir_lang/index.html">dfir_<wbr>lang</a><span class="version">0.14.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module ops</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#constants" title="Constants">Constants</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2><a href="../index.html">In dfir_<wbr>lang::<wbr>graph</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../../index.html">dfir_lang</a>::<wbr><a href="../index.html">graph</a></div><h1>Module <span>ops</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../../src/dfir_lang/graph/ops/mod.rs.html#1-662">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>DFIR’s operators</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.OperatorConstraints.html" title="struct dfir_lang::graph::ops::OperatorConstraints">Operator<wbr>Constraints</a></dt><dd>An instance of this struct represents a single dfir operator.</dd><dt><a class="struct" href="struct.OperatorWriteOutput.html" title="struct dfir_lang::graph::ops::OperatorWriteOutput">Operator<wbr>Write<wbr>Output</a></dt><dd>The code generated and returned by a <a href="struct.OperatorConstraints.html#structfield.write_fn" title="field dfir_lang::graph::ops::OperatorConstraints::write_fn"><code>OperatorConstraints::write_fn</code></a>.</dd><dt><a class="struct" href="struct.WriteContextArgs.html" title="struct dfir_lang::graph::ops::WriteContextArgs">Write<wbr>Context<wbr>Args</a></dt><dd>Context arguments provided to <a href="struct.OperatorConstraints.html#structfield.write_fn" title="field dfir_lang::graph::ops::OperatorConstraints::write_fn"><code>OperatorConstraints::write_fn</code></a>.</dd></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.DelayType.html" title="enum dfir_lang::graph::ops::DelayType">Delay<wbr>Type</a></dt><dd>The delay (soft barrier) type, for each input to an operator if needed.</dd><dt><a class="enum" href="enum.FloType.html" title="enum dfir_lang::graph::ops::FloType">FloType</a></dt><dd>Operator type for Flo semantics.</dd><dt><a class="enum" href="enum.OperatorCategory.html" title="enum dfir_lang::graph::ops::OperatorCategory">Operator<wbr>Category</a></dt><dd>Operator categories, for docs.</dd><dt><a class="enum" href="enum.Persistence.html" title="enum dfir_lang::graph::ops::Persistence">Persistence</a></dt><dd>Persistence lifetimes: <code>'none</code>, <code>'tick</code>, <code>'static</code>, or <code>'mutable</code>.</dd><dt><a class="enum" href="enum.PortListSpec.html" title="enum dfir_lang::graph::ops::PortListSpec">Port<wbr>List<wbr>Spec</a></dt><dd>Specification of the named (or unnamed) ports for an operator’s inputs or outputs.</dd></dl><h2 id="constants" class="section-header">Constants<a href="#constants" class="anchor">§</a></h2><dl class="item-table"><dt><a class="constant" href="constant.IDENTITY_WRITE_FN.html" title="constant dfir_lang::graph::ops::IDENTITY_WRITE_FN">IDENTITY_<wbr>WRITE_<wbr>FN</a></dt><dd><a href="struct.OperatorConstraints.html#structfield.write_fn" title="field dfir_lang::graph::ops::OperatorConstraints::write_fn"><code>OperatorConstraints::write_fn</code></a> for unary identity operators.</dd><dt><a class="constant" href="constant.NULL_WRITE_FN.html" title="constant dfir_lang::graph::ops::NULL_WRITE_FN">NULL_<wbr>WRITE_<wbr>FN</a></dt><dd><a href="struct.OperatorConstraints.html#structfield.write_fn" title="field dfir_lang::graph::ops::OperatorConstraints::write_fn"><code>OperatorConstraints::write_fn</code></a> for the null operator - an operator that ignores all inputs
and produces no output.</dd><dt><a class="constant" href="constant.OPERATORS.html" title="constant dfir_lang::graph::ops::OPERATORS">OPERATORS</a></dt><dd>All DFIR operators.</dd><dt><a class="constant" href="constant.RANGE_0.html" title="constant dfir_lang::graph::ops::RANGE_0">RANGE_0</a></dt><dd>Convenience range: exactly zero.</dd><dt><a class="constant" href="constant.RANGE_1.html" title="constant dfir_lang::graph::ops::RANGE_1">RANGE_1</a></dt><dd>Convenience range: exactly one.</dd><dt><a class="constant" href="constant.RANGE_ANY.html" title="constant dfir_lang::graph::ops::RANGE_ANY">RANGE_<wbr>ANY</a></dt><dd>Convenience range: zero or more (any number).</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><dl class="item-table"><dt><a class="trait" href="trait.RangeTrait.html" title="trait dfir_lang::graph::ops::RangeTrait">Range<wbr>Trait</a></dt><dd>An object-safe version of <a href="https://doc.rust-lang.org/nightly/core/ops/range/trait.RangeBounds.html" title="trait core::ops::range::RangeBounds"><code>RangeBounds</code></a>.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.find_node_op_constraints.html" title="fn dfir_lang::graph::ops::find_node_op_constraints">find_<wbr>node_<wbr>op_<wbr>constraints</a></dt><dd>Find an operator by <a href="../enum.GraphNode.html" title="enum dfir_lang::graph::GraphNode"><code>GraphNode</code></a>.</dd><dt><a class="fn" href="fn.find_op_op_constraints.html" title="fn dfir_lang::graph::ops::find_op_op_constraints">find_<wbr>op_<wbr>op_<wbr>constraints</a></dt><dd>Find an operator by an AST <a href="../../parse/struct.Operator.html" title="struct dfir_lang::parse::Operator"><code>Operator</code></a>.</dd><dt><a class="fn" href="fn.identity_write_iterator_fn.html" title="fn dfir_lang::graph::ops::identity_write_iterator_fn">identity_<wbr>write_<wbr>iterator_<wbr>fn</a></dt><dd>Helper to write the <code>write_iterator</code> portion of <a href="struct.OperatorConstraints.html#structfield.write_fn" title="field dfir_lang::graph::ops::OperatorConstraints::write_fn"><code>OperatorConstraints::write_fn</code></a> output for
unary identity operators.</dd><dt><a class="fn" href="fn.null_write_iterator_fn.html" title="fn dfir_lang::graph::ops::null_write_iterator_fn">null_<wbr>write_<wbr>iterator_<wbr>fn</a></dt><dd>Helper to write the <code>write_iterator</code> portion of <a href="struct.OperatorConstraints.html#structfield.write_fn" title="field dfir_lang::graph::ops::OperatorConstraints::write_fn"><code>OperatorConstraints::write_fn</code></a> output for
the null operator - an operator that ignores all inputs and produces no output.</dd><dt><a class="fn" href="fn.operator_lookup.html" title="fn dfir_lang::graph::ops::operator_lookup">operator_<wbr>lookup</a></dt><dd>Get the operator lookup table, generating it if needed.</dd></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><dl class="item-table"><dt><a class="type" href="type.WriteFn.html" title="type dfir_lang::graph::ops::WriteFn">WriteFn</a></dt><dd>Type alias for <a href="struct.OperatorConstraints.html#structfield.write_fn" title="field dfir_lang::graph::ops::OperatorConstraints::write_fn"><code>OperatorConstraints::write_fn</code></a>’s type.</dd></dl></section></div></main></body></html>