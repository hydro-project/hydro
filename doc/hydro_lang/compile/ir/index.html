<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `ir` mod in crate `hydro_lang`."><title>hydro_lang::compile::ir - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../../static.files/rustdoc-e56847b5.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="hydro_lang" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (f6092f224 2025-09-22)" data-channel="nightly" data-search-js="search-e256b49e.js" data-stringdex-js="stringdex-061df703.js" data-settings-js="settings-c38705f0.js" ><script src="../../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-6dc2a7f3.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module ir</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../hydro_lang/index.html">hydro_<wbr>lang</a><span class="version">0.14.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module ir</a></h2><h3><a href="#modules">Module Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2><a href="../index.html">In hydro_<wbr>lang::<wbr>compile</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../../index.html">hydro_lang</a>::<wbr><a href="../index.html">compile</a></div><h1>Module <span>ir</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../../src/hydro_lang/compile/ir/mod.rs.html#1-3162">Source</a> </span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><dl class="item-table"><dt><a class="mod" href="backtrace/index.html" title="mod hydro_lang::compile::ir::backtrace">backtrace</a></dt><dd>Platform-independent interface for collecting backtraces, used in the Hydro IR to
trace the origin of each node.</dd></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.DebugExpr.html" title="struct hydro_lang::compile::ir::DebugExpr">Debug<wbr>Expr</a></dt><dd>Wrapper that displays only the tokens of a parsed expr.</dd><dt><a class="struct" href="struct.DebugInstantiateFinalized.html" title="struct hydro_lang::compile::ir::DebugInstantiateFinalized">Debug<wbr>Instantiate<wbr>Finalized</a></dt><dt><a class="struct" href="struct.DebugType.html" title="struct hydro_lang::compile::ir::DebugType">Debug<wbr>Type</a></dt><dd>Debug displays the type’s tokens.</dd><dt><a class="struct" href="struct.HydroIrMetadata.html" title="struct hydro_lang::compile::ir::HydroIrMetadata">Hydro<wbr>IrMetadata</a></dt><dt><a class="struct" href="struct.HydroIrOpMetadata.html" title="struct hydro_lang::compile::ir::HydroIrOpMetadata">Hydro<wbr>IrOp<wbr>Metadata</a></dt><dd>Metadata that is specific to the operator itself, rather than its outputs.
This is available on <em>both</em> inner nodes and roots.</dd><dt><a class="struct" href="struct.QMacroSimplifier.html" title="struct hydro_lang::compile::ir::QMacroSimplifier">QMacro<wbr>Simplifier</a></dt><dd>AST visitor that simplifies q! macro expansions</dd><dt><a class="struct" href="struct.TeeNode.html" title="struct hydro_lang::compile::ir::TeeNode">TeeNode</a></dt></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.BuildersOrCallback.html" title="enum hydro_lang::compile::ir::BuildersOrCallback">Builders<wbr>OrCallback</a></dt><dt><a class="enum" href="enum.DebugInstantiate.html" title="enum hydro_lang::compile::ir::DebugInstantiate">Debug<wbr>Instantiate</a></dt><dt><a class="enum" href="enum.HydroNode.html" title="enum hydro_lang::compile::ir::HydroNode">Hydro<wbr>Node</a></dt><dd>An intermediate node in a Hydro graph, which consumes data
from upstream nodes and emits data to downstream nodes.</dd><dt><a class="enum" href="enum.HydroRoot.html" title="enum hydro_lang::compile::ir::HydroRoot">Hydro<wbr>Root</a></dt><dd>An root in a Hydro graph, which is an pipeline that doesn’t emit
any downstream values. Traversals over the dataflow graph and
generating DFIR IR start from roots.</dd><dt><a class="enum" href="enum.HydroSource.html" title="enum hydro_lang::compile::ir::HydroSource">Hydro<wbr>Source</a></dt><dd>A source in a Hydro graph, where data enters the graph.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.dbg_dedup_tee.html" title="fn hydro_lang::compile::ir::dbg_dedup_tee">dbg_<wbr>dedup_<wbr>tee</a></dt><dt><a class="fn" href="fn.deep_clone.html" title="fn hydro_lang::compile::ir::deep_clone">deep_<wbr>clone</a></dt><dt><a class="fn" href="fn.emit.html" title="fn hydro_lang::compile::ir::emit">emit</a></dt><dt><a class="fn" href="fn.transform_bottom_up.html" title="fn hydro_lang::compile::ir::transform_bottom_up">transform_<wbr>bottom_<wbr>up</a></dt><dt><a class="fn" href="fn.traverse_dfir.html" title="fn hydro_lang::compile::ir::traverse_dfir">traverse_<wbr>dfir</a></dt></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><dl class="item-table"><dt><a class="type" href="type.SeenTeeLocations.html" title="type hydro_lang::compile::ir::SeenTeeLocations">Seen<wbr>TeeLocations</a></dt><dt><a class="type" href="type.SeenTees.html" title="type hydro_lang::compile::ir::SeenTees">Seen<wbr>Tees</a></dt></dl></section></div></main></body></html>