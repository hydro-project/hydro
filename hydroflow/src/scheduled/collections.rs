pub struct Single<T>(pub T);

pub struct Iter<I: Iterator>(pub I);
