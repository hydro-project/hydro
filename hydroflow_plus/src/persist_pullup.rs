use std::cell::RefCell;
use std::collections::HashSet;
use std::ops::Deref;

use crate::ir::*;

fn persist_pullup_node<'a>(
    node: &mut HfPlusNode<'a>,
    seen_tees: &mut SeenTees<'a>,
    persist_pulled_tees: &mut HashSet<*const RefCell<HfPlusNode<'a>>>,
) {
    node.transform_children(
        |n, s| persist_pullup_node(n, s, persist_pulled_tees),
        seen_tees,
    );

    match node {
        HfPlusNode::Unpersist(box HfPlusNode::Persist(_)) => {
            if let HfPlusNode::Unpersist(box HfPlusNode::Persist(box behind_persist)) =
                std::mem::replace(node, HfPlusNode::Placeholder)
            {
                *node = behind_persist;
            } else {
                unreachable!()
            }
        }

        HfPlusNode::Delta(box HfPlusNode::Persist(_)) => {
            if let HfPlusNode::Delta(box HfPlusNode::Persist(box behind_persist)) =
                std::mem::replace(node, HfPlusNode::Placeholder)
            {
                *node = behind_persist;
            } else {
                unreachable!()
            }
        }

        HfPlusNode::Tee { inner } => {
            if persist_pulled_tees.contains(&(inner.as_ref() as *const RefCell<HfPlusNode<'a>>)) {
                *node = HfPlusNode::Persist(Box::new(HfPlusNode::Tee {
                    inner: inner.clone(),
                }));
            } else {
                let inner_borrow = inner.borrow();
                if let HfPlusNode::Persist(_) = inner_borrow.deref() {
                    drop(inner_borrow);
                    persist_pulled_tees.insert(inner.as_ref() as *const RefCell<HfPlusNode<'a>>);
                    if let HfPlusNode::Persist(box behind_persist) =
                        inner.replace(HfPlusNode::Placeholder)
                    {
                        *inner.borrow_mut() = behind_persist;
                    } else {
                        unreachable!()
                    }

                    *node = HfPlusNode::Persist(Box::new(HfPlusNode::Tee {
                        inner: inner.clone(),
                    }));
                }
            }
        }

        HfPlusNode::Map {
            f: _,
            input: box HfPlusNode::Persist(_),
        } => {
            if let HfPlusNode::Map {
                f,
                input: box HfPlusNode::Persist(behind_persist),
            } = std::mem::replace(node, HfPlusNode::Placeholder)
            {
                *node = HfPlusNode::Persist(Box::new(HfPlusNode::Map {
                    f,
                    input: behind_persist,
                }));
            } else {
                unreachable!()
            }
        }

        HfPlusNode::FlatMap {
            f: _,
            input: box HfPlusNode::Persist(_),
        } => {
            if let HfPlusNode::FlatMap {
                f,
                input: box HfPlusNode::Persist(behind_persist),
            } = std::mem::replace(node, HfPlusNode::Placeholder)
            {
                *node = HfPlusNode::Persist(Box::new(HfPlusNode::FlatMap {
                    f,
                    input: behind_persist,
                }));
            } else {
                unreachable!()
            }
        }

        HfPlusNode::Filter {
            f: _,
            input: box HfPlusNode::Persist(_),
        } => {
            if let HfPlusNode::Filter {
                f,
                input: box HfPlusNode::Persist(behind_persist),
            } = std::mem::replace(node, HfPlusNode::Placeholder)
            {
                *node = HfPlusNode::Persist(Box::new(HfPlusNode::Filter {
                    f,
                    input: behind_persist,
                }));
            } else {
                unreachable!()
            }
        }

        HfPlusNode::Network {
            input: box HfPlusNode::Persist(_),
            ..
        } => {
            if let HfPlusNode::Network {
                from_location,
                to_location,
                serialize_pipeline,
                instantiate_fn,
                deserialize_pipeline,
                input: box HfPlusNode::Persist(behind_persist),
            } = std::mem::replace(node, HfPlusNode::Placeholder)
            {
                *node = HfPlusNode::Persist(Box::new(HfPlusNode::Network {
                    from_location,
                    to_location,
                    serialize_pipeline,
                    instantiate_fn,
                    deserialize_pipeline,
                    input: behind_persist,
                }));
            } else {
                unreachable!()
            }
        }

        HfPlusNode::Union(box HfPlusNode::Persist(_), box HfPlusNode::Persist(_)) => {
            if let HfPlusNode::Union(
                box HfPlusNode::Persist(left),
                box HfPlusNode::Persist(right),
            ) = std::mem::replace(node, HfPlusNode::Placeholder)
            {
                *node = HfPlusNode::Persist(Box::new(HfPlusNode::Union(left, right)));
            } else {
                unreachable!()
            }
        }

        HfPlusNode::CrossProduct(box HfPlusNode::Persist(_), box HfPlusNode::Persist(_)) => {
            if let HfPlusNode::CrossProduct(
                box HfPlusNode::Persist(left),
                box HfPlusNode::Persist(right),
            ) = std::mem::replace(node, HfPlusNode::Placeholder)
            {
                *node = HfPlusNode::Persist(Box::new(HfPlusNode::Delta(Box::new(
                    HfPlusNode::CrossProduct(
                        Box::new(HfPlusNode::Persist(left)),
                        Box::new(HfPlusNode::Persist(right)),
                    ),
                ))));
            } else {
                unreachable!()
            }
        }

        HfPlusNode::Join(box HfPlusNode::Persist(_), box HfPlusNode::Persist(_)) => {
            if let HfPlusNode::Join(box HfPlusNode::Persist(left), box HfPlusNode::Persist(right)) =
                std::mem::replace(node, HfPlusNode::Placeholder)
            {
                *node =
                    HfPlusNode::Persist(Box::new(HfPlusNode::Delta(Box::new(HfPlusNode::Join(
                        Box::new(HfPlusNode::Persist(left)),
                        Box::new(HfPlusNode::Persist(right)),
                    )))));
            } else {
                unreachable!()
            }
        }

        _ => {}
    }
}

pub fn persist_pullup(ir: Vec<HfPlusLeaf>) -> Vec<HfPlusLeaf> {
    let mut seen_tees = Default::default();
    let mut persist_pulled_tees = Default::default();
    ir.into_iter()
        .map(|l| {
            l.transform_children(
                |n, s| persist_pullup_node(n, s, &mut persist_pulled_tees),
                &mut seen_tees,
            )
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use stageleft::*;

    use crate::deploy::MultiGraph;

    #[test]
    fn persist_pullup_through_map() {
        let flow = crate::builder::FlowBuilder::new();
        let process = flow.process::<()>();

        flow.source_iter(&process, q!(0..10))
            .map(q!(|v| v + 1))
            .for_each(q!(|n| println!("{}", n)));

        let built = flow.finalize();

        insta::assert_debug_snapshot!(built.ir());

        let optimized = built.optimize_with(super::persist_pullup);

        insta::assert_debug_snapshot!(optimized.ir());
        for (id, graph) in optimized.compile_no_network::<MultiGraph>().hydroflow_ir() {
            insta::with_settings!({snapshot_suffix => format!("surface_graph_{id}")}, {
                insta::assert_snapshot!(graph.surface_syntax_string());
            });
        }
    }

    #[test]
    fn persist_pullup_behind_tee() {
        let flow = crate::builder::FlowBuilder::new();
        let process = flow.process::<()>();

        let before_tee = flow.source_iter(&process, q!(0..10)).tick_batch().persist();

        before_tee
            .clone()
            .map(q!(|v| v + 1))
            .all_ticks()
            .for_each(q!(|n| println!("{}", n)));

        before_tee
            .clone()
            .map(q!(|v| v + 1))
            .all_ticks()
            .for_each(q!(|n| println!("{}", n)));

        let built = flow.finalize();

        insta::assert_debug_snapshot!(built.ir());

        let optimized = built.optimize_with(super::persist_pullup);

        insta::assert_debug_snapshot!(optimized.ir());

        for (id, graph) in optimized.compile_no_network::<MultiGraph>().hydroflow_ir() {
            insta::with_settings!({snapshot_suffix => format!("surface_graph_{id}")}, {
                insta::assert_snapshot!(graph.surface_syntax_string());
            });
        }
    }
}
