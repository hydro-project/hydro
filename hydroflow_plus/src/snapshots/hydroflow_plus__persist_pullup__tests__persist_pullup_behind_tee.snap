---
source: hydroflow_plus/src/persist_pullup.rs
expression: built.ir()
---
[
    ForEach {
        f: { use crate :: __staged :: persist_pullup :: tests :: * ; | n | println ! ("{}" , n) },
        input: Unpersist(
            Persist(
                Map {
                    f: { use crate :: __staged :: persist_pullup :: tests :: * ; | v | v + 1 },
                    input: Tee {
                        inner: RefCell {
                            value: Persist(
                                Unpersist(
                                    Persist(
                                        Source {
                                            source: Iter(
                                                { use crate :: __staged :: persist_pullup :: tests :: * ; 0 .. 10 },
                                            ),
                                            location_kind: Process(
                                                0,
                                            ),
                                        },
                                    ),
                                ),
                            ),
                        },
                    },
                },
            ),
        ),
    },
    ForEach {
        f: { use crate :: __staged :: persist_pullup :: tests :: * ; | n | println ! ("{}" , n) },
        input: Unpersist(
            Persist(
                Map {
                    f: { use crate :: __staged :: persist_pullup :: tests :: * ; | v | v + 1 },
                    input: Tee {
                        inner: RefCell {
                            value: Persist(
                                Unpersist(
                                    Persist(
                                        Source {
                                            source: Iter(
                                                { use crate :: __staged :: persist_pullup :: tests :: * ; 0 .. 10 },
                                            ),
                                            location_kind: Process(
                                                0,
                                            ),
                                        },
                                    ),
                                ),
                            ),
                        },
                    },
                },
            ),
        ),
    },
]
